File: app/src/main/java/com/annapol04/munchkin/engine/Match.java
package com.annapol04.munchkin.engine;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.StringRes;
import android.util.Log;
import android.util.Pair;

import com.annapol04.munchkin.R;
import com.annapol04.munchkin.data.EventRepository;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;

import static com.annapol04.munchkin.engine.EngineTest.test;

@Singleton
public class Match {
    protected static final int AMOUNT_OF_HAND_CARDS = 2;
    protected static final int MAX_AMOUNT_OF_HAND_CARDS = 4;

    protected static final String TAG = Match.class.getSimpleName();

    protected Game game;
    protected EventRepository eventRepository;

    protected MutableLiveData<String> log = new MutableLiveData<>();
    protected MutableLiveData<List<Player>> observablePlayers = new MutableLiveData<>();
    protected MutableLiveData<Player> currentPlayer = new MutableLiveData<>();

    protected List<Player> players = new ArrayList<>();
    protected Player current = null;
    protected Player first = null;
    protected final Player myself;
    protected TurnPhase turnPhase;

    protected State state = State.JOINING;
    protected int amountOfPlayers = 0;
    protected int playersRenamed = 0;

    protected boolean named = false;
    protected boolean handCardsDrawn = false;

    protected enum State {
        JOINING,
        NAMING,
        HAND_CARDS,
        STARTED,
    }

    @Inject
    public Match(Game game, @Named("myself") Player myself, EventRepository eventRepository) {
        this.game = game;
        this.myself = myself;
        this.eventRepository = eventRepository;

        observablePlayers.setValue(new ArrayList<>());

        reset();
    }

    protected void reset() {
        players.clear();
        currentPlayer.setValue(null);
        playersRenamed = 0;
        observablePlayers.getValue().clear();
        log.setValue("");

        game.reset();
        myself.reset();

        turnPhase = TurnPhase.IDLE;
        state = State.JOINING;

        named = false;
        handCardsDrawn = false;
    }

    public boolean isMyself(Player player) {
        return player == myself;
    }

    public void log(String message) {
        log.setValue(new StringBuilder()
                .append(log.getValue())
                .append(message)
                .append("\n")
                .toString());
    }

    public void undoLog() {
        log.setValue(log.getValue().replaceAll("\n[^\n]+\n+$", "\n"));
    }

    public LiveData<String> getLog() {
        return log;
    }

    public LiveData<List<Player>> getPlayers() {
        return observablePlayers;
    }

    public int getAmountOfPlayers() {
        return amountOfPlayers;
    }

    public Player getPlayer(int playerNr) {
        return players.get(playerNr - 1);
    }

    public Player getPlayer(Scope scope) {
        return getPlayer(scope.ordinal());
    }

    public LiveData<Player> getCurrentPlayer() {
        return currentPlayer;
    }

    public void start(int amountOfPlayers) {
        reset();

        this.amountOfPlayers = amountOfPlayers;

        eventRepository.push(new Event(Scope.GAME, Action.JOIN_PLAYER, 0, myself.getId()));
    }

    protected Player evaluateHost() {
        Player highestId = null;

        for (Player player : players) {
            if (highestId == null)
                highestId = player;
            else if (player.getId() > highestId.getId())
                highestId = player;
        }

        return highestId;
    }

    protected void specifyPlayerNumbers() {
        myself.setScope(Scope.PLAYER1);

        eventRepository.push(
                new Event(myself.getScope(), Action.ASSIGN_PLAYER_NUMBER, 0, myself.getId()));

        int nr = 2;

        for (Player player : players)
            if (player != myself)
                eventRepository.push(
                        new Event(Scope.fromId(nr++), Action.ASSIGN_PLAYER_NUMBER, 0, player.getId()));
    }

    protected <T> void update(MutableLiveData<T> liveData) {
        liveData.setValue(liveData.getValue());
    }

    public void joinPlayer(int randomNumber) {
        Player player = randomNumber == myself.getId() ?
                myself : new Player(randomNumber, game, eventRepository);

        player.allowToDrawTreasureCards(AMOUNT_OF_HAND_CARDS);

        players.add(player);

        if (players.size() == amountOfPlayers) {
            state = State.NAMING;

            current = evaluateHost();
            currentPlayer.setValue(current);

            first = current;

            if (current == myself) {
                specifyPlayerNumbers();

                named = true;

                namingRound();
            }
        }
    }

    protected void namingRound() {
        eventRepository.push(
                new Event(current.getScope(), Action.NAME_PLAYER, R.string.ev_join_player, myself.getName()),
                new Event(current.getScope(), Action.FINISH_ROUND, 0)
        );
    }

    protected Player nextPlayer(int playerNr) {
        return players.get(playerNr % players.size());
    }

    protected void drawInitialHandcards() {
        game.getRandomTreasureCards(AMOUNT_OF_HAND_CARDS)
                .forEach(card -> {
                    eventRepository.push(
                            new Event(current.getScope(), Action.DRAW_TREASURECARD, 0, card.getId())
                    );           });

        eventRepository.push(
                new Event(current.getScope(), Action.FINISH_ROUND, 0)
        );
    }

    public void finishRound(int playerNr) throws IllegalEngineStateException {
        current = nextPlayer(playerNr);
        currentPlayer.setValue(current);

        switch (state) {
            default:
                throw new IllegalEngineStateException("Illegal state");
            case NAMING:
                if (!named && current == myself) {
                    named = true;

                    namingRound();
                }
                if (named && current == first)
                    state = State.HAND_CARDS;
                break;
            case HAND_CARDS:
                if (!handCardsDrawn && current == myself) {
                    handCardsDrawn = true;

                    Log.e(TAG, "" + handCardsDrawn);
                    drawInitialHandcards();
                }
                if (handCardsDrawn && current == first)
                    state = State.STARTED;
                break;
            case STARTED:
                playRound();
                break;
        }
    }

    protected void emitFinishRound(Scope scope) {
        eventRepository.push(
                new Event(current.getScope(), Action.FINISH_ROUND, 0)
        );
    }

    protected void playRound() throws IllegalEngineStateException {
        if (turnPhase == TurnPhase.IDLE)
            emitEnterTurnPhase(current.getScope(), TurnPhase.EQUIPMENT);
        else
            throw new IllegalEngineStateException("We can start the turn only from \"idle\" turn phase");
    }

    public void assignPlayerNumber(int playerNr, int randomNumber) {
        boolean allScopesAssigned = true;

        for (Player player : players) {
            if (player.getId() == randomNumber)
                player.setScope(Scope.fromId(playerNr));
            if (player.getScope() == null)
                allScopesAssigned = false;
        }

        if (allScopesAssigned)
            players = players.stream()
                    .sorted(Comparator.comparingInt(p -> p.getScope().ordinal()))
                    .collect(Collectors.toList());
    }

    public void namePlayer(int playerNr, String name) {
        players.get(playerNr - 1).rename(name);

        if (++playersRenamed == amountOfPlayers)
            observablePlayers.setValue(players);
    }

    protected void testHost(Player player) throws IllegalEngineStateException {
        if (player != current)
            throw new IllegalEngineStateException("can not play a card in another player's round");
    }


    /****************************************************************************************
     *                      ==== Action implementations ====
     ****************************************************************************************/

    public void emitMessage(Scope scope, @StringRes int message, int integer) {
        eventRepository.push(
                new Event(scope, Action.MESSAGE, message, integer)
        );
    }

    public void emitDrawDoorCard(Scope scope) {
        Card card = game.getRandomDoorCards(1).get(0);

        emitEnterTurnPhase(scope, TurnPhase.KICK_OPEN_THE_DOOR);

        eventRepository.push(
            new Event(scope, Action.DRAW_DOORCARD, R.string.ev_draw_card, card.getId())
        );
    }

    public void drawDoorCard(Scope scope, Card card) throws IllegalEngineStateException {
        testHost(getPlayer(scope));

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR,
                "it's not allowed to draw door cards in \"" + turnPhase + "\" phase!");
        test(game.getDeskCards().getValue().size() == 0,
                "you can not draw a door card from an empty deck!");

        getPlayer(scope).drawDoorCard(card);
    }


    public void emitDrawTreasureCard(Scope scope) {
        Card card = game.getRandomTreasureCards(1).get(0);

        eventRepository.push(
            new Event(scope, Action.DRAW_TREASURECARD, R.string.ev_draw_card, card.getId())
        );
    }

    public void drawTreasureCard(Scope scope, Card card) throws IllegalEngineStateException {
        testHost(getPlayer(scope));

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW || turnPhase == TurnPhase.IDLE,
                "it's not allowed to draw treasure cards in \"" + turnPhase + "\" phase!");
        test(game.getDeskCards().getValue().size() == 0,
                "you can not draw a treasure card from an empty deck!");

        getPlayer(scope).drawTreasureCard(card);

        if (turnPhase == TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW
                && scope == myself.getScope()
                && !myself.isAllowedToDrawTreasureCard())
            emitEnterTurnPhase(scope, TurnPhase.CHARITY);
    }


    public void emitFightMonster(Scope scope) {
        emitEnterTurnPhase(scope, TurnPhase.KICK_OPEN_THE_DOOR_AND_FIGHT);

        eventRepository.push(
                new Event(scope, Action.FIGHT_MONSTER, R.string.ev_fight_monster)
        );
    }

    public void fightMonster(Scope scope) throws IllegalEngineStateException {
        testHost(getPlayer(scope));

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR_AND_FIGHT,
                "monsters can not fought in \"" + turnPhase + "\" phase!");

        Pair<Monster, Integer> result = getPlayer(scope).fightMonster();

        final int cardsToDraw = 1;

        if (scope == myself.getScope()) {
            emitMessage(scope, R.string.player_killed_monster, result.first.getId());
            emitMessage(scope, R.string.player_gets_level, result.second);

            emitEnterTurnPhase(scope, TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW);
            emitMessage(scope, R.string.player_draws_treasure_cards, cardsToDraw);
        }

        getPlayer(scope).allowToDrawTreasureCards(cardsToDraw);
    }


    public void emitRunAway(Scope scope) {
        eventRepository.push(
                new Event(scope, Action.RUN_AWAY, R.string.ev_run_away)
        );
    }

    public void runAway(Scope scope) throws IllegalEngineStateException {
        testHost(getPlayer(scope));

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR_AND_FIGHT,
                "can not run away from monster in \"" + turnPhase + "\" phase!");

        getPlayer(scope).runAway();
    }


    public void emitPlayCard(Scope scope, Card card) {
        eventRepository.push(
                new Event(scope, Action.PLAY_CARD, R.string.ev_play_card, card.getId())
        );
    }


    public void playCard(Scope scope, Card card) throws IllegalEngineStateException {
        testHost(getPlayer(scope));

        test(turnPhase == TurnPhase.EQUIPMENT,
                "can not equip items in \"" + turnPhase + "\" phase!");

        getPlayer(scope).playCard(card);
    }


    protected void emitEnterTurnPhase(Scope scope, TurnPhase phase) {
        eventRepository.push(
                new Event(scope, Action.ENTER_TURN_PHASE, R.string.ev_enter_turn_phase, phase.ordinal())
        );
    }

    public void enterTurnPhase(Scope scope, TurnPhase phase) throws IllegalEngineStateException {
        switch (turnPhase) {
            case IDLE:
                test(phase == TurnPhase.EQUIPMENT,
                        "it is not allowed to enter phase \"" + phase + "\" from \"" + turnPhase + "\"");

                getPlayer(scope).allowToDrawDoorCards(1);
                break;
            case EQUIPMENT:
                test(phase == TurnPhase.KICK_OPEN_THE_DOOR,
                        "it is not allowed to enter phase \"" + phase + "\" from \"" + turnPhase + "\"");
                break;
            case KICK_OPEN_THE_DOOR:
                test(phase == TurnPhase.KICK_OPEN_THE_DOOR_AND_FIGHT,
                        "it is not allowed to enter phase \"" + phase + "\" from \"" + turnPhase + "\"");
                break;
            case KICK_OPEN_THE_DOOR_AND_FIGHT:
                test(phase == TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW,
                        "it is not allowed to enter phase \"" + phase + "\" from \"" + turnPhase + "\"");
                break;
            case KICK_OPEN_THE_DOOR_AND_DRAW:
            case LOOK_FOR_TROUBLE:
            case LOOT_THE_ROOM:
                test(phase == TurnPhase.CHARITY,
                        "it is not allowed to enter phase \"" + phase + "\" from \"" + turnPhase + "\"");

                if (scope == myself.getScope()) {
                    emitEnterTurnPhase(scope, TurnPhase.IDLE);
                    emitFinishRound(scope);
                }
                break;
            case CHARITY:
                break;
        }

        turnPhase = phase;
    }
}

File: app/src/main/java/com/annapol04/munchkin/engine/Match.kt
package com.annapol04.munchkin.engine

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.support.annotation.StringRes

import com.annapol04.munchkin.R
import com.annapol04.munchkin.data.EventRepository

import com.annapol04.munchkin.util.*

import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton
import kotlin.math.E

@Singleton
open class Match @Inject
constructor(protected var game: Game,
            @param:Named("myself") protected val myself: Player,
            protected var eventRepository: EventRepository) {

    protected val log_ by lazy { NonNullMutableLiveData<String>("") }
    val log: NonNullLiveData<String> get() = log_

    protected val players_ by lazy { NonNullMutableLiveData<List<Player>>(emptyList()) }
    val players: NonNullLiveData<List<Player>> get() = players_

    protected val currentPlayer_ by lazy { NonNullMutableLiveData(myself) }
    val currentPlayer: NonNullLiveData<Player> get() = currentPlayer_

    protected val canFinishRound_ by lazy { NonNullMutableLiveData(false) }
    val canFinishRound: NonNullLiveData<Boolean> get() = canFinishRound_

    protected val canStartCombat_ by lazy { NonNullMutableLiveData(false) }
    val canStartCombat: NonNullLiveData<Boolean> get() = canStartCombat_

    protected var current: Player = myself
        set(value) {
            field = value
            currentPlayer_.value = field
        }

    protected var first = myself
    protected var turnPhase = TurnPhase.IDLE
        set(value) {
            field = value

            if (value == TurnPhase.FINISHED)
                canFinishRound_.value = true
            else if (canFinishRound_.value)
                canFinishRound_.value = false

            if (value == TurnPhase.KICK_OPEN_THE_DOOR)
                canStartCombat_.value = true
            else if (canStartCombat_.value)
                canStartCombat_.value = false
        }

    protected var state = State.JOINING
    var amountOfPlayers = 0
        protected set
    protected var playersRenamed = 0

    protected var named = false
    protected var handCardsDrawn = false

    protected enum class State {
        JOINING,
        NAMING,
        HAND_CARDS,
        STARTED
    }

    init {
        reset()
    }

    protected fun reset() {
        players_.value = emptyList()
        playersRenamed = 0
        log_.value = ""

        game.reset()
        myself.reset()

        turnPhase = TurnPhase.IDLE
        state = State.JOINING

        named = false
        handCardsDrawn = false
        current = myself
    }

    fun isMyself(player: Player?): Boolean {
        return player === myself
    }

    fun log(message: String) {
        log_.value = StringBuilder()
                .append(log_.value)
                .append(message)
                .append("\n")
                .toString()
    }

    fun undoLog() {
        log_.value = log_.value.replace("\n[^\n]+\n+$".toRegex(), "\n")
    }

    fun getPlayer(playerNr: Int): Player {
        return players_.value[playerNr - 1]
    }

    fun getPlayer(scope: Scope): Player {
        return getPlayer(scope.ordinal)
    }

    open fun start(amountOfPlayers: Int) {
        reset()

        this.amountOfPlayers = amountOfPlayers

        eventRepository.push(Event(Scope.GAME, Action.JOIN_PLAYER, 0, myself.id))
    }

    protected fun evaluateHost(): Player {
        return players_.value.maxBy { it.id } ?: myself
    }

    protected fun specifyPlayerNumbers() {
        myself.scope = Scope.PLAYER1

        eventRepository.push(
                Event(myself.scope, Action.ASSIGN_PLAYER_NUMBER, 0, myself.id))

        var nr = 2

        for (player in players_.value)
            if (player != myself)
                eventRepository.push(
                        Event(Scope.fromId(nr++), Action.ASSIGN_PLAYER_NUMBER, 0, player.id))
    }

    fun joinPlayer(randomNumber: Int) {
        val player = if (randomNumber == myself.id)
            myself
        else
            Player(randomNumber, game, eventRepository)

        player.allowToDrawTreasureCards(AMOUNT_OF_HAND_CARDS)

        players_.value += player

        if (players_.value.size == amountOfPlayers) {
            state = State.NAMING

            current = evaluateHost()
            first = current

            if (current == myself) {
                specifyPlayerNumbers()

                named = true

                namingRound()
            }
        }
    }

    fun canPlayerFightMonster(player: Player): Boolean {
        return game.deskCards.value.size > 0
                && player.getFightLevel().value!! > (game.deskCards.value[0] as Monster).level

    }

    protected open fun namingRound() {
        eventRepository.push(
                Event(current.scope, Action.NAME_PLAYER, R.string.ev_join_player, myself.getName()),
                Event(current.scope, Action.HAND_OVER_TOKEN, 0, nextPlayer().scope.ordinal)
        )
    }

    protected fun nextPlayer(): Player {
        return players_.value[(players_.value.indexOf(current) + 1) % players_.value.size]
    }

    protected fun drawInitialHandcards() {
        eventRepository.push(game.getRandomTreasureCards(AMOUNT_OF_HAND_CARDS)
                .map{ Event(current.scope, Action.DRAW_TREASURECARD, 0, it.id) })

        emitHandOverToken(current.scope, nextPlayer())
    }

    private fun areAllPlayersNamed(): Boolean {
        return playersRenamed == amountOfPlayers
    }

    private fun areAllInitialCardsDrawn(): Boolean {
        return players_.value.all{ !it.isAllowedToDrawTreasureCard }
    }

    protected fun emitHandOverToken(scope: Scope, player: Player) {
        eventRepository.push(
                Event(current.scope, Action.HAND_OVER_TOKEN, 0, player.scope.ordinal)
        )
    }

    @Throws(IllegalEngineStateException::class)
    open fun handOverToken(scope: Scope, playerNr: Int) {
        if (state === Match.State.STARTED)
            test(canFinishRound_.value,"You can hand over the control after doing the required steps")

        current = getPlayer(playerNr)

        turnPhase = TurnPhase.IDLE

        when (state) {
            Match.State.NAMING -> {
                if (!named && current == myself) {
                    named = true

                    namingRound()
                }
                if (areAllPlayersNamed()) {
                    state = State.HAND_CARDS

                    if (!handCardsDrawn && current == myself) {
                        handCardsDrawn = true

                        drawInitialHandcards()
                    }
                    if (areAllInitialCardsDrawn()) {
                        state = State.STARTED

                        playRound()
                    }
                }
            }
            Match.State.HAND_CARDS -> {
                if (!handCardsDrawn && current == myself) {
                    handCardsDrawn = true
                    drawInitialHandcards()
                }
                if (areAllInitialCardsDrawn()) {
                    state = State.STARTED

                    playRound()
                }
            }
            Match.State.STARTED -> {
                playRound()
            }
            else -> throw IllegalEngineStateException("Illegal state")
        }
    }

    protected fun playRound() {
        turnPhase = TurnPhase.EQUIPMENT
        current.allowToDrawDoorCards(1)
    }

    fun assignPlayerNumber(playerNr: Int, randomNumber: Int) {
        var allScopesAssigned = true

        for (player in players_.value) {
            if (player.id == randomNumber)
                player.scope = Scope.fromId(playerNr)
            if (player.scope === Scope.GAME)
                allScopesAssigned = false
        }

        if (allScopesAssigned)
            players_.value = players_.value.sortedBy{ it.scope.ordinal }
    }

    fun namePlayer(playerNr: Int, name: String) {
        players_.value[playerNr - 1].rename(name)
        ++playersRenamed
    }

    @Throws(IllegalEngineStateException::class)
    protected fun testHost(player: Player) {
        if (player !== current)
            throw IllegalEngineStateException("can not play a card in another player's round")
    }

    fun startCombat() {
        emitMessage(current.scope, R.string.tp_kick_open_the_door_and_fight);
    }


    /****************************************************************************************
     * ==== Action implementations ====
     */

    fun emitMessage(scope: Scope, @StringRes message: Int) {
        eventRepository.push(
                Event(scope, Action.MESSAGE, message)
        )
    }

    fun emitMessage(scope: Scope, @StringRes message: Int, integer: Int) {
        eventRepository.push(
                Event(scope, Action.MESSAGE, message, integer)
        )
    }

    fun emitDrawDoorCard(scope: Scope) {
        val card = game.getRandomDoorCards(1)[0]

        emitMessage(scope, R.string.tp_kick_open_the_door)

        eventRepository.push(
                Event(scope, Action.DRAW_DOORCARD, R.string.ev_draw_card, card.id)
        )
    }

    @Throws(IllegalEngineStateException::class)
    fun drawDoorCard(scope: Scope, card: Card) {
        testHost(getPlayer(scope))

        if (turnPhase == TurnPhase.EQUIPMENT)
            turnPhase = TurnPhase.KICK_OPEN_THE_DOOR

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR,
                "it's not allowed to draw door cards in \"$turnPhase\" phase!")
        test(game.deskCards.value.size == 0,
                "you can not draw a door card from an empty deck!")

        getPlayer(scope).drawDoorCard(card)
    }


    fun emitDrawTreasureCard(scope: Scope) {
        val card = game.getRandomTreasureCards(1)[0]

        eventRepository.push(
                Event(scope, Action.DRAW_TREASURECARD, R.string.ev_draw_card, card.id)
        )
    }

    @Throws(IllegalEngineStateException::class)
    fun drawTreasureCard(scope: Scope, card: Card) {
        testHost(getPlayer(scope))

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW || turnPhase == TurnPhase.IDLE,
                "it's not allowed to draw treasure cards in \"$turnPhase\" phase!")
        test(game.deskCards.value.size == 0,
                "you can not draw a treasure card from an empty deck!")

        getPlayer(scope).drawTreasureCard(card)

        if (turnPhase == TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW
        && !current.isAllowedToDrawTreasureCard) {
            current.limitHandCards(max = MAX_AMOUNT_OF_HAND_CARDS)

            startCharitingOrFinish();
        }
    }

    private fun startCharitingOrFinish() {
        if (current.isAllowedToDropCard) {
            if (current == myself) {
                if (myself.handCards.value.size > MAX_AMOUNT_OF_HAND_CARDS)
                    emitMessage(current.scope, R.string.tp_charity)
            }

            turnPhase = TurnPhase.CHARITY
        } else
            turnPhase = TurnPhase.FINISHED
    }


    fun emitFightMonster(scope: Scope) {
        eventRepository.push(
                Event(scope, Action.FIGHT_MONSTER, R.string.ev_fight_monster)
        )
    }

    @Throws(IllegalEngineStateException::class)
    fun fightMonster(scope: Scope) {
        testHost(getPlayer(scope))

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR,
                "monsters can not fought in \"$turnPhase\" phase!")

        val result = getPlayer(scope).fightMonster()

        val cardsToDraw = 1

        if (scope == myself.scope) {
            emitMessage(scope, R.string.player_killed_monster, result.first.id)
            emitMessage(scope, R.string.player_gets_level, result.second)
            emitMessage(scope, R.string.tp_kick_open_the_door_and_draw)
            emitMessage(scope, R.string.player_draws_treasure_cards, cardsToDraw)
        }

        getPlayer(scope).allowToDrawTreasureCards(cardsToDraw)

        turnPhase = TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW
    }


    fun emitRunAway(scope: Scope) {
        eventRepository.push(
                Event(scope, Action.RUN_AWAY, R.string.ev_run_away)
        )
    }

    @Throws(IllegalEngineStateException::class)
    fun runAway(scope: Scope) {
        testHost(getPlayer(scope))

        test(turnPhase == TurnPhase.KICK_OPEN_THE_DOOR,
                "can not run away from monster in \"$turnPhase\" phase!")

        getPlayer(scope).runAway()

        startCharitingOrFinish()
    }


    fun emitPlayCard(scope: Scope, card: Card) {
        eventRepository.push(
                Event(scope, Action.PLAY_CARD, R.string.ev_play_card, card.id)
        )
    }

    @Throws(IllegalEngineStateException::class)
    fun playCard(scope: Scope, card: Card) {
        testHost(getPlayer(scope))

        test(turnPhase == TurnPhase.EQUIPMENT,
                "can not equip items in \"$turnPhase\" phase!")

        getPlayer(scope).playCard(card)
    }


    fun emitDropCard(scope: Scope, card: Card) {
        eventRepository.push(
                Event(scope, Action.DROP_CARD, R.string.ev_drop_card, card.id)
        )
    }

    fun dropCard(scope: Scope, card: Card) {
        testHost(getPlayer(scope))

        test(turnPhase == TurnPhase.CHARITY,
                "can not drop cards in \"$turnPhase\" phase!")

        getPlayer(scope).dropCard(card)

        if (turnPhase == TurnPhase.CHARITY && !current.isAllowedToDropCard)
            turnPhase = TurnPhase.FINISHED
    }


   /* private fun emitEnterTurnPhase(scope: Scope, phase: TurnPhase) {
        eventRepository.push(
                Event(scope, Action.ENTER_TURN_PHASE, R.string.ev_enter_turn_phase, phase.ordinal)
        )
    }*/

 /*   @Throws(IllegalEngineStateException::class)
    fun enterTurnPhase(scope: Scope, phase: TurnPhase) {
        when (turnPhase) {
            TurnPhase.IDLE -> {
                test(phase == TurnPhase.EQUIPMENT,
                        "it is not allowed to enter phase \"$phase\" from \"$turnPhase\"")


            }
            TurnPhase.EQUIPMENT -> test(phase == TurnPhase.KICK_OPEN_THE_DOOR,
                    "it is not allowed to enter phase \"$phase\" from \"$turnPhase\"")
            TurnPhase.KICK_OPEN_THE_DOOR -> test(phase == TurnPhase.KICK_OPEN_THE_DOOR_AND_FIGHT,
                    "it is not allowed to enter phase \"$phase\" from \"$turnPhase\"")
            TurnPhase.KICK_OPEN_THE_DOOR_AND_FIGHT -> test(phase == TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW,
                    "it is not allowed to enter phase \"$phase\" from \"$turnPhase\"")
            TurnPhase.KICK_OPEN_THE_DOOR_AND_DRAW, TurnPhase.LOOK_FOR_TROUBLE, TurnPhase.LOOT_THE_ROOM -> {
                test(phase == TurnPhase.CHARITY,
                        "it is not allowed to enter phase \"$phase\" from \"$turnPhase\"")

                if (scope == myself.scope) {
       //             emitEnterTurnPhase(scope, TurnPhase.IDLE)
                    emitHandOverToken(scope, nextPlayer())
                }
            }
            TurnPhase.CHARITY -> {
            }
        }

        turnPhase = phase

        if (turnPhase == TurnPhase.CHARITY)
            canFinishRound_.value = true
        else if (canFinishRound_.value)
            canFinishRound_.value = false

        if (turnPhase == TurnPhase.KICK_OPEN_THE_DOOR)
            canStartCombat_.value = true
        else if (canStartCombat_.value)
            canStartCombat_.value = false
    }
*/
    fun finishRound() {
        emitHandOverToken(current.scope, nextPlayer())
    }

    companion object {
        protected val AMOUNT_OF_HAND_CARDS = 2
        protected val MAX_AMOUNT_OF_HAND_CARDS = 4

        protected val TAG = Match::class.java.simpleName
    }
}